### DISCUSSION TITLE
Understanding TM/TC with regard to replacing F' packets with Space Packet

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1228] [Q&A]
Hi,

I would like to replace the F' packet with CCSDS Space Packet. This seems like it could be quite a big job, particularly since I am completely new to F'.

Can anyone offer end-to-end information on TM and TC processing in F', particularly a data-flow diagram?

*Edit:* I am on F' 2.1.0

Many thanks,

Jack

**Comment by JackNWhite** [comment]
From looking at the F' framing protocol, my understanding is that frames follow this spec:

- start word (0xDEADBEEF): 32b
- data length: 32b
- packet_type: 32b (unless it's unknown)
- payload: unspecified length
- hash: 32b/256b

Is this correct? How does a frame relate to a packet? Is there a written specification of the F' packet format?

**Reply by LeStarch** [reply]
The above looks correct.   Note: packet type is always there, "unknown" types are set to file buffers.

We are working on a PR that describes the whole ground interface, which is where I'd suggest you start. https://github.com/nasa/fprime/pull/1202

**Reply by LeStarch** [reply]
That PR should be the same for v2.1.0 and v3.0.0.

**Reply by JackNWhite** [reply]
Thank you

**Reply by JackNWhite** [reply]
Thanks again, @LeStarch. I've read your guide, but I am not clear on what aspect of communication is represented by the frame, and how it ties in with the Fw::ComPacket and Fw::CmdPacket types.

The framing protocol doesn't seem to have any way of accessing information about the source of the data, or any other metadata. What is the framing protocol supposed to represent?

Will you include information on CmdPacket and ComPacket in your guide? It would be most helpful if you could provide a field-by-field diagram of F' packets/frames, and a data flow diagram of how they are constructed within the flight software.

For example, this Space Packet diagram:
![spp](https://user-images.githubusercontent.com/10062543/150110791-2470f943-d7a1-4a32-a6c8-88acdfcf155a.png)

Surely the packet protocol is important in getting a ground system to understand your communication?

**Reply by rawinza555** [reply]
Out of curiosity, what's the secondary header flag is for? I assume that it could help when there are more than two types of data since the packet type is only 1 bit long. 

In your opinion, if ones want to adapt CCSDS packet to the F', should we modify framer and deframer so that the packet comes out like this or should we add like a component between comm and uplink/downlink and have that component append stuff to the F' packet so that entire F' packet fits in user data field?

**Reply by JackNWhite** [reply]
The secondary header flag is set if your packets have a secondary header, which is optional. European institutional missions typically use the ECSS PUS standard secondary headers, which does have a 16-bit packet type field, although it's split into service and subtype.

I did look at FramingProtocol, DeframingProtocol and FprimeProtocol first for reading and writing packets, but how would it work? The Space Packet APID is purposefully of mission design, so you could easily match this up with components' base ID. The offsets are a different matter though, because they are reused between events and channels, which are both really just telemetry.

In any case, neither the base ID nor the offsets appear to be available to those classes. At least, not that I can work out. I really have difficulty with OOP, and F' seems to be quite a complicated example.

I'm not sure what you mean by "between comm and uplink/downlink". I can't see any components called those things.

I would really like a diagram of the TM/TC system so I can work out how the data flows between the up- and downlinks and the components. Surely there must be a diagram somewhere!

**Reply by timcanham** [reply]
I've been putting thought into how we might implement CCSDS for F Prime. WE have it on our roadmap to do at some point because there is a lot of interest. My thoughts for APID: APID would identify the type of data that an F prime deployment is producing. On the uplink side, it would be: 1) A command, 2) a file (or piece of a file) (CFDP is a big bite to think about for now). On the downlink side, it would be: 1) Events, 2) Channelized telemetry, 3) packetized telemetry (upcoming feature), 4) [Data Products](https://github.com/nasa/fprime/discussions/1154), and 5) files (same APID as uplink). I thought about using the secondary header for things like channel/event ID, but we may want to stack those in a given packet, so I was just going to make the IDs part of the user data. A component ID as APID is interesting, but I think it breaks the model that the ground system would like be generic, not knowing about components per se. Once the ground system sees the data type via the APID, it can then walk through the user data, parsing it with the dictionary.

**Reply by timcanham** [reply]
So in light of that, we could write a Framer/Deframer for F Prime that takes the packets and reformats them in CCSDS, and a corresponding plug-in for the ground system.

**Reply by ebenezerarunkumar** [reply]
Adding CCSDS Functionality would be great @timcanham ! I'm involved with a few Fprime projects at Georgia Tech and I know the teams here would greatly benefit from this. Please let us know if/how we can help!

**Reply by JackNWhite** [reply]
Typically, the APID is, at least partly, used to identify the flight software application or functional area in which the packet originates or for which it is destined. That's a very mission-orientated scheme.

Also, there seems little real difference between an event packet, a channel packet and a packet packet.

It would be better to retain the flexibility and let the mission decide on APIDs. One could have a mission-specific converter service to translate between IDs internal to the deployment and APIDs.  The APID would need to be sent to `eventLogger`, `chanTlm` or `pktTlm`. You could specify your APID, or APID subcomponent in your XML/FPP file.

I'm just wary, from my own current situation, that you might want TM/TC for an F' subsystem to form part of a larger system using the same APID scheme.

**Reply by JackNWhite** [reply]
With reference to the PR you mentioned, @LeStarch, that guide did not show the addition of the packet ID to the packet - it's not written/read by the framing protocol, but in TlmPacket. Could this be added to ground interface guide?

**Reply by JackNWhite** [reply]
It now looks to me (based on FprimeProtocol and TlmPacket) as if the structure of an F' packet is:
![fprime_packet-1](https://user-images.githubusercontent.com/10062543/151530110-77261002-77af-40cf-8bfd-8c0c001d8996.png)

How does this match other people's knowledge?

**Comment by JackNWhite** [comment]
I'm thinking that the chanTlm/pktTlm/eventLogger/Framer/FramingProtocol system might stand to be reworked a little.

What is the difference between an event and a channel from the comms perspective? Why are there separate queues for channels and events? Is there a mechanism for bumping high-priority packets up the queue?

From which point is the serialisation of the packets called? I'm struggling with the design of the system here, coming from a world of C. Do you build up the packet metadata in TlmPacketizer, TlmChan and ActiveLogger, and then serialise the metadata, along with the payload, in Framer/FramingProtocol? Or is the payload already serialised by the time it reaches Framer/FramingProtocol? I'm confused by OOP, so excuse my ignorance!

I'm just thinking that, if you want to do Space Packet, then all the packet data and metadata must already be available by the time it is queued in, for example, TlmChan. Why, then are packets not fully serialised before they are queued?

Here is the Space Packet primary header:
![image](https://user-images.githubusercontent.com/10062543/150969235-dc19f36b-e3a7-4a12-822e-b1c1f760a8be.png)

... Bear in mind that there might also be secondary header, itself containing further metadata which would have to be queued.

What about a structure more like this?
![Possible FPrime TM](https://user-images.githubusercontent.com/10062543/150970850-90d0099e-a949-41a9-8a17-ca65e2d974a8.svg)
